src/workflow-requests.ts

```diff
@@ -20,79 +20,103 @@ import type {
 } from "./types";
 import { StepTypes } from "./types";
 import type { WorkflowLogger } from "./logger";
-import { FlowControl, PublishRequest, QstashError } from "@upstash/qstash";
+import { FlowControl, PublishBatchRequest, PublishRequest, QstashError } from "@upstash/qstash";
 import { getSteps } from "./client/utils";
 import { getHeaders } from "./qstash/headers";
+import { PublishToUrlResponse } from "@upstash/qstash";
 
-export const triggerFirstInvocation = async <TInitialPayload>({
-  workflowContext,
-  useJSONContent,
-  telemetry,
-  debug,
-  invokeCount,
-  delay,
-}: {
+type TriggerFirstInvocationParams<TInitialPayload> = {
   workflowContext: WorkflowContext<TInitialPayload>;
   useJSONContent?: boolean;
   telemetry?: Telemetry;
   debug?: WorkflowLogger;
   invokeCount?: number;
   delay?: PublishRequest["delay"];
-}): Promise<Ok<"success" | "workflow-run-already-exists", never> | Err<never, Error>> => {
-  const { headers } = getHeaders({
-    initHeaderValue: "true",
-    workflowConfig: {
-      workflowRunId: workflowContext.workflowRunId,
-      workflowUrl: workflowContext.url,
-      failureUrl: workflowContext.failureUrl,
-      retries: workflowContext.retries,
-      telemetry,
-      flowControl: workflowContext.flowControl,
-      useJSONContent: useJSONContent ?? false,
-    },
-    invokeCount: invokeCount ?? 0,
-    userHeaders: workflowContext.headers,
-  });
+};
 
-  // QStash doesn't forward content-type when passed in `upstash-forward-content-type`
-  // so we need to pass it in the headers
-  if (workflowContext.headers.get("content-type")) {
-    headers["content-type"] = workflowContext.headers.get("content-type")!;
-  }
+export const triggerFirstInvocation = async <TInitialPayload>(
+  params:
+    | TriggerFirstInvocationParams<TInitialPayload>
+    | TriggerFirstInvocationParams<TInitialPayload>[]
+): Promise<Ok<"success" | "workflow-run-already-exists", never> | Err<never, Error>> => {
+  const firstInvocationParams = Array.isArray(params) ? params : [params];
+  const workflowContextClient = firstInvocationParams[0].workflowContext.qstashClient;
 
-  if (useJSONContent) {
-    headers["content-type"] = "application/json";
-  }
+  const invocationBatch = firstInvocationParams.map((invocation) => {
```

Comment:
you can destructure invocation like this to simplify the code below: `firstInvocationParams.map(({ workflowContext, useJSONContent, delay })`